File: scala/euler004.scala
object euler004 {
    def isPalindrome(n : Int) : Boolean = {
        val str = n.toString()
        val rev = str.reverse.toString
        str == rev
    }

    def main(args : Array[String]) : Unit = {
        val numbers = for {
            a <- (100 to 999).reverse;
            b <- (100 to a).reverse
            prod = a * b
            if isPalindrome(prod)
        } yield prod
        val result = numbers.reduceLeft(_.max(_))
        println(result)
    }
}

File: java/euler004.java
class euler004 {
    private static String reverse(String str) {
        char[] chars = str.toCharArray();
        for (int i = 0; i < chars.length/2; i++) {
            char tmp = chars[i];
            chars[i] = chars[chars.length-i-1];
            chars[chars.length-i-1] = tmp;
        }
        StringBuffer buff = new StringBuffer();
        for (int i = 0; i < chars.length; i++) {
            buff.append(chars[i]);
        }
        return buff.toString();
    }

    private static boolean isPalindrome(int num) {
        String str = "" + num;
        String rev = reverse(str);
        return str.equals(rev);
    }

    public static void main(String[] args) {
        int max = -1;
        for (int a = 999; a >= 100; a--) {
            for (int b = a; b >= 100; b--) {
                int num = a * b;
                if ((num > max) && isPalindrome(num)) {
                    max = num;
                }
            }
        }
        System.out.println(max);
    }
}
